pkg/fleet-manager/backup_restore_migrate_shared.go

```diff
@@ -0,0 +1,336 @@
+func generateVeleroResourceName(clusterName, creatorKind, creatorNamespace, creatorName string) string {
+	return clusterName + "-" + creatorKind + "-" + creatorNamespace + "-" + creatorName
+}
+
+// MostRecentCompletedBackup returns the most recent backup that's completed from a list of backups.
+// origin from https://github.com/vmware-tanzu/velero/blob/release-1.12/pkg/controller/restore_controller.go
+func MostRecentCompletedBackup(backups []velerov1.Backup) velerov1.Backup {
+	sort.Slice(backups, func(i, j int) bool {
+		var iStartTime, jStartTime time.Time
+		if backups[i].Status.StartTimestamp != nil {
+			iStartTime = backups[i].Status.StartTimestamp.Time
+		}
+		if backups[j].Status.StartTimestamp != nil {
+			jStartTime = backups[j].Status.StartTimestamp.Time
+		}
+		return iStartTime.After(jStartTime)
+	})
+
+	for _, backup := range backups {
+		if backup.Status.Phase == velerov1.BackupPhaseCompleted {
+			return backup
+		}
+	}
+
+	return velerov1.Backup{}
+}
+
+// GetCronInterval return the cron interval of a cron expression。
+func GetCronInterval(cronExpr string) (time.Duration, error) {
+	parser := cron.NewParser(cron.Minute | cron.Hour | cron.Dom | cron.Month | cron.Dow)
+	schedule, err := parser.Parse(cronExpr)
+	if err != nil {
+		return 0, err
+	}
+
+	now := time.Now()
+	nextRun := schedule.Next(now)
+	nextNextRun := schedule.Next(nextRun)
+
+	// Adding a 30-second delay to avoid timing issues.
+	// Without this delay, we risk checking just before a new backup starts,
+	// seeing the previous backup's "completed" status, and missing the new one.
+	const delay = 30 * time.Second
+	interval := nextNextRun.Sub(nextRun) + delay
+
+	return interval, nil
+}
+
+// getResourceFromClusterClient retrieves a specific Kubernetes resource from the provided cluster.
+func getResourceFromClusterClient(ctx context.Context, name, namespace string, clusterAccess fleetCluster, obj client.Object) error {
+	clusterClient := clusterAccess.client.CtrlRuntimeClient()
+
+	resourceKey := types.NamespacedName{
+		Name:      name,
+		Namespace: namespace,
+	}
+	return clusterClient.Get(ctx, resourceKey, obj)
+}
+
+// listResourcesFromClusterClient retrieves resources from a cluster based on the provided namespace and label.
+func listResourcesFromClusterClient(ctx context.Context, namespace string, labelKey string, labelValue string, clusterAccess fleetCluster, objList client.ObjectList) error {
+	// Create the cluster client
+	clusterClient := clusterAccess.client.CtrlRuntimeClient()
+	// Create the label selector
+	labelSelector := labels.Set(map[string]string{labelKey: labelValue}).AsSelector()
+
+	// List the resources
+	opts := &client.ListOptions{
+		Namespace:     namespace,
+		LabelSelector: labelSelector,
+	}
+	return clusterClient.List(ctx, objList, opts)
+}
+
+// parallelProcess runs the provided tasks concurrently and collects any errors.
+func parallelProcess(tasks []func() error) []error {
```

Comment:
to be accurate, it is concurrent.

```suggestion
func runTasks(tasks []func() error) []error {
```

Reply:
this fucn is no more needed, removed it.